### ISSUE TITLE
An error occurred when training with [sedf_st] using 'ut extract_hypno' 

### ISSUE THREAD

**Issue by ljyljy** [perslev/U-Time#23]
Hello, after using the 'ut extract_hypno' command, I processed the original data from sedf_sc and sedf_st separately, and then performed training. 
For the sedf_sc data, there is no error in the training for the time being.
But for sedf_st data, an error is reported:
"ValueError("Init times and durations do not match.")"

Complete command：
ut extract_hypno --file_regex 'datasets/sedf_st/ST*/*Hypnogram.edf' \
--out_dir 'datasets/processed/sedf_st/' 
--overwrite 

error screenshot:
[Dataset: sedf_ st/TRAIN] Loading 30/30 SLeepStudy objects...
[WARNING] Quality control for sample ' ST7081J0' affected 1/967 epochs in channel G
[WARNING] Quality control for sample ST7081J0' affected 25/967 epochs in channel 1
[WARNING] Quality control for sample ' ST7242J0' affected 6/940 epochs in channel 0
[WARNING] Quality control for sample ' ST7242J0' affected 27/940 epochs in channel 1
[ WARNING] Quality control for sample ST7102J0' affected 9/986 epochs in channel 0
[ WARNING] Quality control for sample ST7102J0' affected 19/986 epochs in channel 1
[WARNING] Quality control for sample ' ST7021J0' affected 4/920 epochs in channel 0
[WARNING] Quality control for sample ' ST7021J0' affected 37/920 epochs in channel 1
[WARNING] Quality control for sample ' ST7011J0' affected 42/ 1092 epochs in channel 0
[WARNING] Quality control for sample ' ST7011J0' affected 49/ 1092 epochs in channel 1
[WARNING] Quality control for sample ' ST7112J0' affected 1/986 epochs in channel 0
[WARNING] Quality control for sample ' ST7112J0' affected 22/986 epochs in channel 1
[WARNING] Quality control for sample 'ST7122J0' affected 13/969 epochs in channel 0
[WARNING] Quality control for sample ' ST7122J0' affected 58/969 epochs in channel 1
[WARNING] Quality control for sample ' ST7201J0' affected 1/924 epochs in channel G
[WARNING] Quality control for sample ' ST7201J0' affected 7/924 epochs in channel 1
[ WARNING] Quality control for sample ST7041J0' affected 4/ 1008 epochs in channel 0
[WARNING] Quality control for sample ' ST7041J0' affected 166/ 1008 epochs in channel 1
T raceback (most recent call last) :
File” /export/jy/U-Time/utime/ dataset/sleep_ study/sleep_ study.py", line 404， in load
self. Load()
File "/export/jy/U-Time/utime/ dataset/sleep study/sleep study.py", line 342, in_ load
S ample_ rate=header["sample_ rate"])
File " /export/Ljy/U-Time/utime/io/high_ Level_ file loaders.py", Line 106, in load hypnogram
sample_ rate=sample_ rate)
File "/export/ljy/U- Time/utime/ io/hypnogram/hyp_ extractors.py", line 144, in extract hyp_ data
ann_ to class-annotation dict
File "/export/jy/U-Time/utime/hypnogram/utils.py", line 284，in sparse hypnogram from ids_ format 
period Length_ sec=period Length sec )
File "/export/Ljy/U-Time/ut ime/hypnogram/hypnograms.py", Line 51，in init
raise ValueError("Init times and durations do not match.")
ValueErrore Init times and durations do not match.

Is anything wrong with the training process or the command input or the project codes?
THX a lot!!!

**Comment by perslev** [comment]
Hi!

I just had a look into this, and I experience the exact same issue on two samples. It occurs because of 'gaps' in the original hypnogram EDF files, where sections of the night's sleep had to annotations. This situation was not correctly accounted for in the current state of hypnogram extraction script.

I have added some small changes to the codebase and `extract_hypno.py` script. Specifically, you may now run:

`ut extract_hypno --file_regex '[LOCAL_PATH]/sleep-edf-extended/ST*/*Hypnogram.edf' --out_dir '[LOCAL_PATH]/processed/sedf-st/' --fill_blanks 'Sleep stage ?'`

_(see https://github.com/perslev/U-Time/blob/master/resources/usleep_dataset_prep/sedf-st.md)_

... where the `--fill_blanks` flag is new and allows to fill such empty sections of the hypnogram with some value. Above, I assign the 'unknown' label to those sections. The specific string 'Sleep stage ?' is just the specific notation for such used for the SEDF datasets, which ensures that the U-Time framework know that the section should not be considered when computing the loss function during training  (see e.g., https://github.com/perslev/U-Time/blob/master/utime/bin/defaults/usleep/dataset_configurations/sedf_st.yaml#L36).

After running the above command, you should re-run `ut preprocess` and then `ut train` should work. Otherwise, please let me know.

Cheers,
Mathias

**Comment by perslev** [comment]
Fixed as of 04d0e925e7da6812f6ded2167738eb4bc5a83261.

**Comment by ljyljy** [comment]
According to your reply and code adjustment, the training has been successfully started!

THX again! ^O^
